Added find_related_containers

kelby-shelton opened the pull request:
Added custom function that takes a container input and match criteria and produces all related containers.

kelby-shelton committed:
Added find_related_containers

Added custom function that takes a container input and match criteria and produces all related containers.

philroyer-phantom commented on custom_functions/find_related_containers.py:
```diff
@@ -0,0 +1,128 @@
+            container_dictionary[str(k)] = list(set(v))
+            
+            # If any of the containers contain more than the minimum match count request that container detail.
+            if len(container_dictionary[str(k)]) > minimum_match_count:
```
I think this should be greater than or equal to minimum_match_count. If a number barely meets the minimum threshold it is still big enough to count.

kelby-shelton replied:
Adjusting in next commit to be >=

philroyer-phantom commented on custom_functions/find_related_containers.json:
```diff
@@ -0,0 +1,78 @@
+            "contains_type": [
+                "*"
+            ],
+            "description": "A cef value to search on such as artifact:*.cef.destinationAddress or to search all indicators in container use: *",
```
I think this is a bit confusing because usually we give an example value that is the literal value being passed in, but your example is the data path which will get converted to a literal value before running. It might be a bit more clear if it said

> An indicator value to search on, such as a file hash or IP address. To search on all indicator values in the container, use "*".

kelby-shelton replied:
Made adjustments to description

kelby-shelton committed:
Implemented feedback

Adjusted > to >= and adjusted description of value_list

kelby-shelton committed:
Dropped artifact_count

Dropped artifact count and changed comment above that section as this was previous functionality that was no longer needed.

kelby-shelton committed:
Made significant changes

- Added several performance increases
- Added several debug statements.
- Added alternate collection means of containers for results over 600.

philroyer-phantom commented on custom_functions/find_related_containers.py:
```diff
@@ -0,0 +1,189 @@
+            response_data = phantom.requests.get(url, verify=False).json()
+            
+            # Populate an indicator dictionary where the original ids are the dictionary keys and the                     
+            # associated continers are the values
```
/s/continers/containers

philroyer-phantom commented on custom_functions/find_related_containers.py:
```diff
@@ -0,0 +1,189 @@
+                
+        # If the container number is greater than 600, then its faster to grab all containers
+        if container_number >= 600:
+            phantom.debug("Number of related containers > 100. Fetching all container data")
```
debug statement should say 600

philroyer-phantom commented on custom_functions/find_related_containers.py:
```diff
@@ -0,0 +1,189 @@
+            phantom.debug("Number of related containers > 100. Fetching all container data")
+            # Convert status to id
+            status_id = None
+            if isinstance(filter_out_status, str):
```
Just for organization purposes, could you please move this container status lookup to the top where you are handling the other similar input validations?

philroyer-phantom replied:
Also, please convert the status name to lowercase before putting it in the query. I am seeing now that status names are represented inconsistently in the UI. They are title case like "Closed" on the investigation page, all caps in the Admin page, and on the backend it seems they are always all lower case. In queries they need to be lower case.

philroyer-phantom commented on custom_functions/find_related_containers.py:
```diff
@@ -0,0 +1,189 @@
+            if isinstance(filter_out_status, str):
+                url = phantom.build_phantom_rest_url('container_status') + f'?_filter_name="{filter_out_status}"'
+                response = phantom.requests.get(uri=url, verify=False).json()
+                if response['count'] > 0:
```
If the status is None or an invalid status name is used, please put a debug statement like:

"The status name was not found, so no containers will be excluded based on their status"

philroyer-phantom commented on custom_functions/find_related_containers.json:
```diff
@@ -0,0 +1,78 @@
+        },
+        {
+            "contains_type": [],
+            "description": "Filters out any containers with this status",
```
Please specify that the status can be a name or status_id.

philroyer-phantom commented on custom_functions/find_related_containers.py:
```diff
@@ -0,0 +1,189 @@
+                    in_case = response_data['in_case']
+
+                    # Build final output
+                    if status != filter_out_status and str(k) != str(current_container):
```
At this point status is the name but filter_out_status could be either the name or the ID. I think in the input processing you should assign filter_out_status_name and filter_out_status_id so you can use the correct one down here.

philroyer-phantom reviewed (changes requested):
Overall it looks very good but there are a handful of minor changes I suggested in the review.
